src/shogun/lib/auto_initialiser.h

```diff
@@ -29,36 +31,107 @@ namespace shogun
 
 			~GammaFeatureNumberInit() override = default;
 
-			Any operator()() override
+			Any operator()() final
 			{
-				REQUIRE(m_kernel != nullptr, "m_kernel is not pointing to a CKernel object");
+				REQUIRE(
+				    m_kernel != nullptr,
+				    "m_kernel is not pointing to a CKernel object");
 
 				auto lhs = m_kernel->get_lhs();
 				switch (lhs->get_feature_class())
 				{
-					case EFeatureClass::C_DENSE:
-					case EFeatureClass::C_SPARSE:
+				case EFeatureClass::C_DENSE:
+				{
+					auto dense_features = (CDenseFeatures<float64_t>*)lhs;
+					return make_any(
+					    1.0 /
+					    static_cast<float64_t>(
+					        dense_features->get_dim_feature_space()) *
+					    dense_features->std(false)[0]);
+				}
+				case EFeatureClass::C_SPARSE:
+				{
+					auto dot_features = (CDotFeatures*)lhs;
+					return make_any(
+					    1.0 / (static_cast<float64_t>(
+					               dot_features->get_dim_feature_space()) *
+					           dot_features->get_std(false)[0]));
+				}
+				default:
+					auto dot_features = (CDotFeatures*)lhs;
+					return make_any(
+					    1.0 / static_cast<float64_t>(
+					              dot_features->get_dim_feature_space()));
+				}
+			}
+
+		private:
+			CKernel* m_kernel;
+			SG_DELETE_COPY_AND_ASSIGN(GammaFeatureNumberInit);
+		};
+
+		class GaussianWidthAutoInit : public AutoInit
+		{
+			static const char* const kName;
+			static const char* const kDescription;
+
+		public:
+			explicit GaussianWidthAutoInit(CKernel* kernel)
+			    : AutoInit(kName, kDescription), m_kernel(kernel)
+			{
+			}
+
+			~GaussianWidthAutoInit() override = default;
+
+			Any operator()() final
+			{
+				REQUIRE(
+				    m_kernel != nullptr,
+				    "m_kernel is not pointing to a CKernel object");
+				auto lhs = m_kernel->get_lhs();
+				auto rhs = m_kernel->get_rhs();
+				switch (lhs->get_feature_class())
+				{
+				case EFeatureClass::C_DENSE:
+				{
+					auto distance_kernel = CEuclideanDistance();
+					distance_kernel.init(lhs, rhs);
+					auto pdist = distance_kernel.get_distance_matrix();
+					auto result = SGVector<float64_t>(
```

Comment:
do we need to copy this... i mean currently you already rely on sgvector being on cpu memory so just then take the eigen map vector of it...

Reply:
What do you mean? I still need to get everything from one of the triangles. The triangular view from Eigen just returns a matrix with 0s in the lower/upper triangle, I still need the non zero values?

Reply:
technically might would be possible to compute the median of upper/lower triangular values without copying them. But I am not so sure that is worth it?

Reply:
However, I think what I would like to see is some analogue to the scipy way of computing pairwise distances, with a function that returns the distances as a vector, and then an "expansion" function. This way you can call the vector function here and then sort that directly, and you save yourself from ever storing the other half of the matrix. This also solves Viktor's comment above.

Reply:
I guess we could add a function in CDistance that does that? And then add a static function that  can do the conversion to square form?

Reply:
and the new class member function would do this https://stackoverflow.com/a/13079806 ?

Reply:
Yes. In fact the method for the full matrix would use the condensed form and expand ir

Reply:
do all the distances in shogun satisfy the triangle inequality and are symmetric? i.e. d(A,B) == d(B,A)

Reply:
nvm, they do, the base class has this written :D